apps/yggdrasil_web/web/controllers/character_controller.ex

```diff
@@ -3,24 +3,40 @@ defmodule YggdrasilWeb.CharacterController do
   use Guardian.Phoenix.Controller
 
   alias Yggdrasil.Character
+  alias YggdrasilWeb.EnsurePermission
+
+  plug EnsurePermission, [character: [:read]] when action in [:index, :show]
+  plug EnsurePermission, [character: [:write]] when action in [:create, :delete]
 
   def index(conn, _params, user, _claims) do
-    user_id = user.id
-    chars = Repo.all from c in Character,
-                    where: c.user_id == ^user_id,
-                    select: c
+    query = if admin?(conn) do
+      Character
+    else
+      from c in Character,
+      where: c.user_id == ^user.id,
+      select: c
+    end
+
+    chars = Repo.all query
 
     render conn, :show, data: chars
   end
 
   def show(conn, %{"char_id" => char_id}, user, _claims) do
-    user_id = user.id
 
     # returns nil if not found
     # need to sort out what we want here.
-    char = Repo.one from c in Character,
-                    where: c.id == ^char_id and c.user_id == ^user_id,
-                    select: c
+    query = if admin?(conn) do
+      from c in Character,
+      where: c.id == ^char_id,
+      select: c
+    else
+      from c in Character,
+      where: c.id == ^char_id and c.user_id == ^user.id,
```

Comment:
Do you mean `Character.visible_to(user)` to hide the 'and c.user_id == ^user.id' or also hides the permission check?

Certainly can be composed, holding off on doing too much in this branch but it can also I think be reduced to 
```elixir
Character
|> where(id: char_id)
|> Character.visible_to(user)
```

Also need to sort out how to handle filters in the index function which I think that can be composed as well. Just holding off for now till I can put a little more thought into it.